tests/Unit/NumericalAlgorithms/LinearOperators/Test_CoefficientTransforms.cpp

```diff
@@ -0,0 +1,169 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/IndexIterator.hpp"
+#include "DataStructures/ModalVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Mesh.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+// These tests generate a function `u_expected` from a linear superposition of
+// the basis functions, which are then transformed to spectral space
+// (u_ceoff) and the coefficients compared with the coefficients set in physical
+// space. The functions are then transformed back (u_back) and compared to the
+// original function, `u_expected`.
+namespace {
+template <Spectral::Basis Basis, Spectral::Quadrature Quadrature>
+void test_1d() noexcept {
+  CAPTURE(Basis);
+  CAPTURE(Quadrature);
+  // Start at 1st order so we are independent of the minimum number of
+  // coefficients.
+  for (size_t order = 1; order < Spectral::maximum_number_of_points<Basis>;
+       ++order) {
+    CAPTURE(order);
+    const Mesh<1> mesh(order + 1, Basis, Quadrature);
+    const DataVector& x = Spectral::collocation_points(mesh);
+    const auto helper = [&mesh, order, &x ](const size_t second_term) noexcept {
+      DataVector u_expected =
+          Spectral::compute_basis_function_value<Basis>(order, x);
+      if (second_term != 1000) {
+        u_expected +=
+            Spectral::compute_basis_function_value<Basis>(second_term, x);
+      }
+      const ModalVector u_coeffs = to_modal_coefficients(u_expected, mesh);
+      const DataVector u_back = to_nodal_coefficients(u_coeffs, mesh);
+      for (size_t i = 0; i <= order; ++i) {
+        if (i == order or i == second_term) {
+          CHECK(u_coeffs[i] == approx(1.0));
+        } else {
+          CHECK(u_coeffs[i] == approx(0.0));
+        }
+      }
+      CHECK_ITERABLE_APPROX(u_expected, u_back);
+    };
+    helper(1000);
+    helper(order / 2);
+    if (order > 4) {
+      helper(order / 3);
+    }
+  }
+}
+
+template <Spectral::Basis Basis, Spectral::Quadrature Quadrature>
+void test_2d() noexcept {
+  // Start at one higher order so we can drop one order in the y-direction.
+  for (size_t order = Spectral::minimum_number_of_points<Basis, Quadrature> + 1;
+       order < Spectral::maximum_number_of_points<Basis>; ++order) {
+    CAPTURE(order);
+    const Mesh<2> mesh({{order + 1, order}}, Basis, Quadrature);
+    const auto logical_coords = logical_coordinates(mesh);
+    const auto helper =
+        [&logical_coords, &mesh, order ](const size_t second_term) noexcept {
+      DataVector u_expected(mesh.number_of_grid_points());
+      for (size_t s = 0; s < mesh.number_of_grid_points(); ++s) {
```

Comment:
Can you assign `u_expected` without the for loop? Like this: `u_expected = Spectral::compute_basis_function_value<Basis>(order, get<0>(logical_coords)) * [same with 0 -> 1];`

If this works, also do this for the "second term" term, and for the 3D test.